modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java

```diff
@@ -428,16 +438,29 @@ static QueryBuilder parseQueryBuilder(DocumentParserContext context) {
 
     static void createQueryBuilderField(
         Version indexVersion,
+        TransportVersion clusterTransportVersion,
         BinaryFieldMapper qbField,
         QueryBuilder queryBuilder,
         DocumentParserContext context
     ) throws IOException {
-        try (ByteArrayOutputStream stream = new ByteArrayOutputStream()) {
-            try (OutputStreamStreamOutput out = new OutputStreamStreamOutput(stream)) {
-                out.setTransportVersion(indexVersion.transportVersion);
-                out.writeNamedWriteable(queryBuilder);
-                qbField.indexValue(context, stream.toByteArray());
+        try (
+            ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            OutputStreamStreamOutput out = new OutputStreamStreamOutput(stream)
+        ) {
+            if (indexVersion.before(Version.V_8_8_0)) {
+                // just use the index version directly
+                // there's a direct mapping from Version to TransportVersion before 8.8.0
+                out.setTransportVersion(TransportVersion.fromId(indexVersion.id));
+            } else {
+                // write the version id to the stream first
+                // NOTE: if needed, at some later point this can be replaced with a specific version number
+                // if the format needs to change again
+                TransportVersion.writeVersion(clusterTransportVersion, out);
```

Comment:
Not really, in that we already have the index version so we know what format we're using.

* Index version <8.8.0 - use matching transport version, written directly
* Index version >=8.8.0 - actual transport version written to stream first

Transport version is just a single vint, and that is highly unlikely to change. If it ever does, we can then change the interpretation of the existing vint field to use it as a separate stream version (vint < 10,523,675 - its transport version, vint == 0xffffff - new format). Or we can piggy-back on the index version again, if that changes at the same time.

Reply:
Perhaps this has more to do with transport version in general, but it seems unnecessary to try to compress these versions, because they are effectively un-compressable (the current transport version already requires 4 bytes as a vint). If we were to write a fixed 4 byte integer, I would be less concerned about the code for reading/writing changing.

Reply:
That's fair - I've changed it to a normal int

Reply:
...however, we can't easily do that, as 8.7 is being released, with transportversion being a vint